### PR TITLE
[Feature] Implement ACRN and Focal Loss

### PR DESCRIPTION
**Pull request by kennymckormick**

### CODE REVIEW

File: configs/detection/acrn/slowfast_acrn_kinetics_pretrained_r50_8x8x1_cosine_10e_ava22_rgb.py [hunk 1]
```diff
@@ -0,0 +1,172 @@
...
+                pos_fraction=1,
+                neg_pos_ub=-1,
+                add_gt_as_proposals=True),
+            pos_weight=1.0,
+            debug=False)),
+    test_cfg=dict(rcnn=dict(action_thr=0.002)))
+
+dataset_type = 'AVADataset'
+data_root = '/mnt/lustre21/DATAshare2/duanhaodong/ava/frames'
+anno_root = '/mnt/lustre21/DATAshare2/duanhaodong/ava/annotations'
```
Comment (dreamerlin) [inline configs/detection/acrn/slowfast_acrn_kinetics_pretrained_r50_8x8x1_cosine_10e_ava22_rgb.py:65]: These two lines should be changed

File: configs/detection/acrn/slowfast_acrn_kinetics_pretrained_r50_8x8x1_cosine_10e_ava22_rgb.py [hunk 2]
```diff
@@ -0,0 +1,172 @@
...
+
+label_file = f'{anno_root}/ava_action_list_v2.2_for_activitynet_2019.pbtxt'
+
+proposal_file_train = f'{anno_root}/train_fair.pkl'
+proposal_file_val = f'{anno_root}/val_fair.pkl'
+
+mc_cfg = dict(
+    server_list_cfg='/mnt/lustre/share/memcached_client/server_list.conf',
+    client_cfg='/mnt/lustre/share/memcached_client/client.conf',
+    sys_path='/mnt/lustre/share/pymc/py3')
```
Comment (dreamerlin) [inline configs/detection/acrn/slowfast_acrn_kinetics_pretrained_r50_8x8x1_cosine_10e_ava22_rgb.py:81]: remove this

File: mmaction/models/heads/bbox_head.py [hunk 3]
```diff
@@ -38,6 +38,8 @@ def __init__(
             spatial_pool_type='max',
             in_channels=2048,
             # The first class is reserved, to classify bbox as pos / neg
+            focal_gamma=0.,
+            focal_alpha=1.,
```
Comment (dreamerlin) [inline mmaction/models/heads/bbox_head.py:42]: need to add docstring

**Review by dreamerlin — CHANGES_REQUESTED** [review]

**Comment by dreamerlin** [comment]
Sth in all config files should be removed.

File: mmaction/models/heads/misc_head.py [hunk 4]
```diff
@@ -0,0 +1,127 @@
...
+    from mmdet.models.builder import SHARED_HEADS as MMDET_SHARED_HEADS
+    mmdet_imported = True
+except (ImportError, ModuleNotFoundError):
+    mmdet_imported = False
+
+# Note: All these heads take 5D Tensors as input (N, C, T, H, W)
+
+
+class ACRNHead(nn.Module):
+    """ACRN Head: Tile + 1x1 convolution + 3x3 convolution.
```
Comment (congee524) [inline mmaction/models/heads/misc_head.py:16]: add paper link

**Comment by congee524** [comment]
Why we put acrn_head in misc_head.py

File: mmaction/models/heads/bbox_head.py [hunk 5]
```diff
@@ -172,7 +177,12 @@ def loss(self,
             labels = labels[pos_inds]
 
             bce_loss = F.binary_cross_entropy_with_logits
-            losses['loss_action_cls'] = bce_loss(cls_score, labels)
+
+            loss = bce_loss(cls_score, labels, reduction='none')
+            pt = torch.exp(-loss)
+            F_loss = self.focal_alpha * (1 - pt)**self.focal_gamma * loss
+            losses['loss_action_cls'] = torch.mean(F_loss)
```
Comment (dreamerlin) [inline mmaction/models/heads/bbox_head.py:184]: How about moving this part as `class FocalLoss` and using it by build_loss(xxx)?

Comment (congee524) [reply mmaction/models/heads/bbox_head.py:184]: like BaseHead

File: mmaction/models/heads/misc_head.py [hunk 6]
```diff
@@ -0,0 +1,127 @@
...
+
+        super().__init__()
+        self.in_channels = in_channels
+        self.out_channels = out_channels
+        self.stride = stride
+        self.num_convs = num_convs
+        self.conv_cfg = conv_cfg
+        self.norm_cfg = norm_cfg
+        self.act_cfg = act_cfg
+        self.pooling = nn.AdaptiveMaxPool3d(1)
```
Comment (dreamerlin) [inline mmaction/models/heads/misc_head.py:48]: `self.pooling` -> `self.max_pool`

**Comment by kennymckormick** [comment]
> Why we put acrn_head in misc_head.py

I will add a series of shared_heads for feature fusion. Currently, I only committed ACRNHead since it's the most successful one.

**Comment by dreamerlin** [comment]
> > Why we put acrn_head in misc_head.py
> 
> I will add a series of shared_heads for feature fusion. Currently, I only committed ACRNHead since it's the most successful one.

I guess what he means is that why not name it as `acrn_head`

**Comment by kennymckormick** [comment]
> > > Why we put acrn_head in misc_head.py
> > 
> > 
> > I will add a series of shared_heads for feature fusion. Currently, I only committed ACRNHead since it's the most successful one.
> 
> I guess what he means is that why not name it as `acrn_head`

I know. Since I will add a series of heads in the future, I plan to put all of them in the misc_head.py to avoid creating a new file for each individual.

File: mmaction/models/heads/bbox_head.py (hunk 5 above)
Comment (kennymckormick) [reply mmaction/models/heads/bbox_head.py:184]: I think we can make FocalLoss a separate module when there is evidence that it can lead to performance gain on other tasks. For now, it is only used in the training of AVA.

**Review by dreamerlin — APPROVED** [review]
